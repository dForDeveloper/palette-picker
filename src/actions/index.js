export const setColors = (colors) => ({
  type: 'SET_COLORS',
  colors
});

export const setLockedColors = (lockedColors) => ({
  type: 'SET_LOCKED_COLORS',
  lockedColors
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const setPalettes = (id, palettes) => ({
  type: 'SET_PALETTES',
  id,
  palettes
});

export const setModal = (isDisplayed, modalType = null, text = null) => ({
  type: 'SET_MODAL',
  isDisplayed,
  modalType,
  text
});