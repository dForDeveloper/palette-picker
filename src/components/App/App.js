import React, { Component } from 'react';

export class App extends Component {
  constructor() {
    super();
    this.state = {
      colors: []
    };
  }

  componentDidMount() {
    this.generateColors();
  }

  generateColors = () => {
    const colors = []
    for (let i = 0; i < 5; i++) {
      let color = '#';
      for (let j = 0; j < 3; j++) {
        const num = Math.floor(Math.random() * 256);
        color += num < 16 ? `0${num.toString(16)}` : num.toString(16);
      }
      colors.push(color);
    }
    this.setState({ colors });
  }

  renderColors = () => {
    return this.state.colors.map(color => {
      return (
        <section
          className="main--color-section"
          key={color}
          style={{ backgroundColor: color }}
        >
        </section>
      );
    });
  }

  render() {
    return (
      <div className="App">
        <h1 className="h1">Palette Picker</h1>
        <button onClick={this.generateColors} className="App--button">
          Generate Palette
        </button>
        <main className="main">{this.renderColors()}</main>
      </div>
    );
  }
}