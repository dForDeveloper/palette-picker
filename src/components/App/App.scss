.App {
  height: 100vh;
}

.h1 {
  height: 55px;
  font-size: 2rem;
  text-align: center;
  margin: 0;
  line-height: 3rem;
  color: #0facf3;
  font-weight: 900;
}

.header {
  position: relative;
}

.header--button {
  height: 35px;
  background-color: #ffffff;
  border: none;
  outline: none;
  cursor: pointer;
  width: 160px;
  margin: 10px auto;
  display: block;
  color: #333333;
  font-size: 1.2rem;

  &:hover {
    color: #0facf3;
  }
}

.header--menu {
  position: absolute;
  top: 65px;
  right: 12px;
  cursor: pointer;
  width: 35px;
  height: 35px;
  background-image: url('../icons/menu.svg');
  
  &:hover {
    background-image: url('../icons/menuhover.svg');
  }
}

.main {
  display: grid;
  height: calc(100vh - 110px);
}

.main--container {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
}

.section {
  width: 100%;
  height: 100%;
  cursor: pointer;
  display: grid;
}

.section--menu {
  width: 100%;
  background-color: #0b0b0e;
  color: #9facbd;
}

@media screen and (max-width: 958px) {
  .main--container {
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: repeat(5, 1fr);
  }
}