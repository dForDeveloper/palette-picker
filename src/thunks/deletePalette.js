import { setError, toggleLoading, setModal, removePalette } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deletePalette = (projectID, paletteID) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');
      await fetchData(`/api/v1/palettes/${paletteID}`, options);
      dispatch(removePalette(projectID, paletteID));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting palette: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}