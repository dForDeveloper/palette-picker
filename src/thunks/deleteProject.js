import { setError, toggleLoading, removeProject, setModal } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');  
      await fetchData(`/api/v1/projects/${id}`, options);
      dispatch(removeProject(id));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting project: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}