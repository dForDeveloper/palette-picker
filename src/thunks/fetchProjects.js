import { setProjects, toggleLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchProjects = () => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const projects = await fetchData('/api/v1/projects');
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError('Error fetch projects: ' + error.message));
    }
    dispatch(toggleLoading(false))
  }
}