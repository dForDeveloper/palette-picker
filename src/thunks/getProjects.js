import { setProjects, toggleLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const getProjects = () => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const projects = await fetchData('/api/v1/projects');
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError('Error getting projects: ' + error.message));
    }
    dispatch(toggleLoading(false))
  }
}