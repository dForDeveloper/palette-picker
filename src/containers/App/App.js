import React, { Component } from 'react';
import { connect } from 'react-redux';
import ColorSection from '../ColorSection/ColorSection';
import SaveModal from '../SaveModal/SaveModal';
import EditModal from '../EditModal/EditModal';
import Menu from '../../containers/Menu/Menu';
import { getProjects } from '../../thunks/getProjects';
import PropTypes from 'prop-types';
import { setColors, setModal } from '../../actions';

export class App extends Component {
  state = {
    menuDisplayed: false
  };

  componentDidMount() {
    this.generateColors();
    this.props.getProjects();
  }

  generateColors = () => {
    const { colors, lockedColors } = this.props;
    const newColors = [];
    for (let i = 0; i < 5; i++) {
      if (lockedColors.includes(i)) {
        newColors.push(colors[i]);
      } else {
        const newColor = '#' + Math.random().toString(16).slice(2, 8);
        newColors.push(newColor);
      }
    }
    this.props.setColors(newColors);
  };

  renderColors = () => {
    return this.props.colors.map((color, index) => {
      return (
        <ColorSection
          key={color}
          color={color}
          index={index}
          toggleLock={this.toggleLock}
        />
      );
    });
  };

  toggleMenu = () => {
    this.setState({ menuDisplayed: !this.state.menuDisplayed });
  };
  
  closeModals = () => {
    this.props.setModal(false);
  };

  render() {
    const { menuDisplayed } = this.state;
    const { modal } = this.props;
    const { modalType } = modal;
    const gridTemplateColumns = menuDisplayed ? '1fr 248px' : '1fr';
    return (
      <div className="App">
        <h1 className="App--h1">palette picker</h1>
        <header>
          <button onClick={this.generateColors} className="header--button">
            generate palette
          </button>
          <div className="header--menu" onClick={this.toggleMenu} />
        </header>
        <main className="main" style={{ gridTemplateColumns }}>
          <div className="main--div">{this.renderColors()}</div>
          {menuDisplayed && <Menu key="menu" />}
          {modalType === 'save' && <SaveModal />}
          {(modalType === 'project' || modalType === 'palette') && 
            <EditModal {...modal} />}
        </main>
        {modal.isDisplayed && (
          <div className="App--overlay" onClick={this.closeModals} />
        )}
      </div>
    );
  }
}

export const mapStateToProps = (state) => ({
  colors: state.colors,
  lockedColors: state.lockedColors,
  projects: state.projects,
  modal: state.modal
});

export const mapDispatchToProps = (dispatch) => ({
  setColors: (colors) => dispatch(setColors(colors)),
  getProjects: () => dispatch(getProjects()),
  setModal: (isDisplayed) => dispatch(setModal(isDisplayed))
});

export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  colors: PropTypes.array,
  lockedColors: PropTypes.array,
  projects: PropTypes.array,
  modal: PropTypes.object,
  setColors: PropTypes.func,
  setModal: PropTypes.func,
  getProjects: PropTypes.func
};
