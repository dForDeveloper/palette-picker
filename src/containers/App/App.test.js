import React from 'react';
import { shallow } from 'enzyme';
import { App, mapStateToProps, mapDispatchToProps } from './App';
import { setColors, setModal } from '../../actions';
import { getProjects } from '../../thunks/getProjects';

jest.mock('../../thunks/getProjects');

describe('App', () => {
  let wrapper;
  const mockProps = {
    colors: ['#aabbcc', '#bbccdd', '#ccddee', '#778899', '#112233'],
    lockedColors: [],
    projects: [],
    modal: { modalType: null },
    setColors: jest.fn(),
    setModal: jest.fn(),
    getProjects: jest.fn()
  };

  beforeEach(() => {
    wrapper = shallow(<App {...mockProps} />);
  });

  describe('render', () => {
    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should match the snapshot when menuDisplayed is true', () => {
      wrapper.setState({ menuDisplayed: true });
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should match the snapshot when modalType is save', () => {
      const mockModal = {
        isDisplayed: true,
        modalType: 'save',
        currentName: null,
        projectID: null,
        paletteID: null
      }
      wrapper = shallow(<App {...mockProps} modal={mockModal} />);
      expect(wrapper).toMatchSnapshot();
    });
    
    it('should match the snapshot when modalType is palette', () => {
      const mockModal = {
        isDisplayed: true,
        modalType: 'palette',
        currentName: 'my palette',
        projectID: 1,
        paletteID: 1
      }
      wrapper = shallow(<App {...mockProps} modal={mockModal} />);
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should call generateColors when generate palette is clicked', () => {
      jest.spyOn(wrapper.instance(), 'generateColors');
      wrapper.instance().forceUpdate();
      wrapper.find('.header--button').simulate('click');
      expect(wrapper.instance().generateColors).toHaveBeenCalled();
    });
    
    it('should call toggleMenu when the menu button is clicked', () => {
      jest.spyOn(wrapper.instance(), 'toggleMenu');
      wrapper.instance().forceUpdate();
      wrapper.find('.header--menu').simulate('click');
      expect(wrapper.instance().toggleMenu).toHaveBeenCalled();
    });
  
    it('should call closeModals when the overlay is clicked', () => {
      const mockModal = {
        isDisplayed: true,
        modalType: 'palette',
        currentName: 'my palette',
        projectID: 1,
        paletteID: 1
      }
      wrapper = shallow(<App {...mockProps} modal={mockModal} />);
      jest.spyOn(wrapper.instance(), 'closeModals');
      wrapper.instance().forceUpdate();
      wrapper.find('.App--overlay').simulate('click');
      expect(wrapper.instance().closeModals).toHaveBeenCalled();
    });
  });

  describe('componentDidMount', () => {
    it('should call generateColors', () => {
      jest.spyOn(wrapper.instance(), 'generateColors');
      wrapper.instance().forceUpdate();
      wrapper.instance().componentDidMount();
      expect(wrapper.instance().generateColors).toHaveBeenCalled();
    });

    it('should call getProjects', () => {
      wrapper.instance().componentDidMount();
      expect(mockProps.getProjects).toHaveBeenCalled();
    });
  });

  describe('generateColors', () => {
    it('should call setColors', () => {
      wrapper.instance().generateColors();
      expect(mockProps.setColors).toHaveBeenCalled();
    });
  });
  
  describe('renderColors', () => {
    it('should return an array of five ColorSections', () => {
      const result = wrapper.instance().renderColors();
      expect(result).toHaveLength(5);
    });
  });

  describe('toggleMenu', () => {
    it('should toggle menuDisplayed in state', () => {
      expect(wrapper.state('menuDisplayed')).toEqual(false);
      wrapper.instance().toggleMenu();
      expect(wrapper.state('menuDisplayed')).toEqual(true);
    });
  });

  describe('closeModals', () => {
    it('should call setModal(false)', () => {
      wrapper.instance().closeModals();
      expect(mockProps.setModal).toHaveBeenCalledWith(false);
    });
  });
});

describe('mapStateToProps', () => {
  it('should return a props object with the correct properties', () => {
    const mockState = {
      colors: ['#abcdef'],
      lockedColors: [0],
      projects: [{ id: 1, name: 'my project' }],
      modal: { isDisplayed: false },
      extraProperty: true
    };
    const expected = {
      colors: ['#abcdef'],
      lockedColors: [0],
      projects: [{ id: 1, name: 'my project' }],
      modal: { isDisplayed: false }
    };
    const result = mapStateToProps(mockState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should dispatch setColors', () => {
    const expected = setColors(['#abcdef']);
    mappedProps.setColors(['#abcdef']);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should dispatch setModal', () => {
    const expected = setModal(false);
    mappedProps.setModal(false);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should dispatch getProjects', () => {
    const expected = getProjects();
    mappedProps.getProjects();
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});