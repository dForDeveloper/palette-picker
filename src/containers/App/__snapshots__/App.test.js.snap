// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App render should match the snapshot 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    palette picker
  </h1>
  <header>
    <button
      className="header--button"
      onClick={[Function]}
    >
      generate palette
    </button>
    <div
      className="header--menu"
      onClick={[Function]}
    />
  </header>
  <main
    className="main"
    style={
      Object {
        "gridTemplateColumns": "1fr",
      }
    }
  >
    <div
      className="main--div"
    >
      <Connect(ColorSection)
        color="#aabbcc"
        index={0}
        key="#aabbcc"
      />
      <Connect(ColorSection)
        color="#bbccdd"
        index={1}
        key="#bbccdd"
      />
      <Connect(ColorSection)
        color="#ccddee"
        index={2}
        key="#ccddee"
      />
      <Connect(ColorSection)
        color="#778899"
        index={3}
        key="#778899"
      />
      <Connect(ColorSection)
        color="#112233"
        index={4}
        key="#112233"
      />
    </div>
  </main>
</div>
`;

exports[`App render should match the snapshot when menuDisplayed is true 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    palette picker
  </h1>
  <header>
    <button
      className="header--button"
      onClick={[Function]}
    >
      generate palette
    </button>
    <div
      className="header--menu"
      onClick={[Function]}
    />
  </header>
  <main
    className="main"
    style={
      Object {
        "gridTemplateColumns": "1fr 248px",
      }
    }
  >
    <div
      className="main--div"
    >
      <Connect(ColorSection)
        color="#aabbcc"
        index={0}
        key="#aabbcc"
      />
      <Connect(ColorSection)
        color="#bbccdd"
        index={1}
        key="#bbccdd"
      />
      <Connect(ColorSection)
        color="#ccddee"
        index={2}
        key="#ccddee"
      />
      <Connect(ColorSection)
        color="#778899"
        index={3}
        key="#778899"
      />
      <Connect(ColorSection)
        color="#112233"
        index={4}
        key="#112233"
      />
    </div>
    <Connect(Menu)
      key="menu"
    />
  </main>
</div>
`;

exports[`App render should match the snapshot when modalType is palette 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    palette picker
  </h1>
  <header>
    <button
      className="header--button"
      onClick={[Function]}
    >
      generate palette
    </button>
    <div
      className="header--menu"
      onClick={[Function]}
    />
  </header>
  <main
    className="main"
    style={
      Object {
        "gridTemplateColumns": "1fr",
      }
    }
  >
    <div
      className="main--div"
    >
      <Connect(ColorSection)
        color="#aabbcc"
        index={0}
        key="#aabbcc"
      />
      <Connect(ColorSection)
        color="#bbccdd"
        index={1}
        key="#bbccdd"
      />
      <Connect(ColorSection)
        color="#ccddee"
        index={2}
        key="#ccddee"
      />
      <Connect(ColorSection)
        color="#778899"
        index={3}
        key="#778899"
      />
      <Connect(ColorSection)
        color="#112233"
        index={4}
        key="#112233"
      />
    </div>
    <Connect(EditModal)
      currentName="my palette"
      isDisplayed={true}
      modalType="palette"
      paletteID={1}
      projectID={1}
    />
  </main>
  <div
    className="App--overlay"
    onClick={[Function]}
  />
</div>
`;

exports[`App render should match the snapshot when modalType is save 1`] = `
<div
  className="App"
>
  <h1
    className="App--h1"
  >
    palette picker
  </h1>
  <header>
    <button
      className="header--button"
      onClick={[Function]}
    >
      generate palette
    </button>
    <div
      className="header--menu"
      onClick={[Function]}
    />
  </header>
  <main
    className="main"
    style={
      Object {
        "gridTemplateColumns": "1fr",
      }
    }
  >
    <div
      className="main--div"
    >
      <Connect(ColorSection)
        color="#aabbcc"
        index={0}
        key="#aabbcc"
      />
      <Connect(ColorSection)
        color="#bbccdd"
        index={1}
        key="#bbccdd"
      />
      <Connect(ColorSection)
        color="#ccddee"
        index={2}
        key="#ccddee"
      />
      <Connect(ColorSection)
        color="#778899"
        index={3}
        key="#778899"
      />
      <Connect(ColorSection)
        color="#112233"
        index={4}
        key="#112233"
      />
    </div>
    <Connect(Modal) />
  </main>
  <div
    className="App--overlay"
    onClick={[Function]}
  />
</div>
`;
